lib/ykpiv.c

```diff
@@ -282,14 +341,17 @@ ykpiv_rc ykpiv_transfer_data(ykpiv_state *state, const unsigned char *templ,
     memcpy(apdu.st.data, in_ptr, this_size);
     res = send_data(state, &apdu, data, &recv_len, sw);
     if(res != YKPIV_OK) {
+      _end_transaction(state);
       return res;
     } else if(*sw != SW_SUCCESS && *sw >> 8 != 0x61) {
+      _end_transaction(state);
```

Comment:
The cases where we do _end_transaction() and then return YKPIV_OK should probably do return(_end_transaction()) no?